hi-bi/nodejs2023Q2-service
# Devlogauth

# Home Library Service

Task: [Home Library Service: Part 3](https://github.com/AlreadyBored/nodejs-assignments/blob/main/assignments/logging-error-authentication-authorization/assignment.md)
Deadline: 2023-08-22 03:00 (UTC+3)
Preliminary assessment: 340 / 340

# Basic Scope
## 1) Logging & Error Handling:
 - [x] **+20** Custom `LoggingService` is implemented and used for logging
 - [x] +**20** Custom `Exception Filter` is implemented and used for handling exceptions during request processing
 - [x] **+20** Logging for request (of at least `url`, `query parameters`, `body`) and response with `status code` is implemented.
 - [x] **+20** Error handling is implemented including sending response with an appropriate `http status` code and errors logging.
 - [x] **+10** Error handling and logging is implemented for `uncaughtException` event.
 - [x] **+10** Error handling and logging is implemented for `unhandledRejection` event.
## 2) Authentication and Authorization:
- [x] **+30** Route `/auth/signup` implemented correctly, related logic is divided between controller and corresponding service
- [x] **+30** Route `/auth/login` has been implemented, related logic is divided between controller and corresponding service
- [x] **+10** `User` `password` saved into database as hash
- [x] **+20** Access Token is implemented, `JWT` payload contains `userId` and `login`, secret key is saved in `.env`.
- [x] **+40** Authentication is required for the access to all routes except `/auth/signup`, `/auth/login`, `/doc` and `/`.
- [x] **+10** Separate module is implemented **within application scope** to check that all requests to all routes except mentioned above contain required JWT token
# Advanced Scope
## 1) Logging & Error Handling:
- [x] **+20** Logs are written to a file.
- [x] **+10** Logs files are rotated with size.
- [x] **+10** Add environment variable to specify max file size.
- [x] **+10** Error logs are written to a separate file (either only to a separate file or in addition to logging into a common file).
- [x] **+20** Add environment variable to specify logging level and corresponding functionality.
Logs with configured level to be registered as well as other higher priority levels. For example if you set level 2, all messages with levels 0, 1 and 2 should be logged. You should use Nest.js logging levels.
## 2) Authentication and Authorization:
- [x] **+30** Route `/auth/refresh` implemented correctly, related logic is divided between controller and corresponding service

## Downloading

```
git clone -b devlogauth https://github.com/hi-bi/nodejs2023Q2-service.git
cd ./nodejs2023Q2-service
```

## Installing

The scripts should be executed under a user with the necessary permissions (sudo ...). 

- Before execution it is necessary to create an `.env` file based on `.env.example` and set values of environment variables in it. 

The application includes the Swagger api, available at http://localhost:${PORT}/doc

- Run the script to initially install and run the containers: 
```
docker compose up
```
<b>Important!</b>
Before testing, you must additionally manually refresh the data model. Additions were made for refreshToken operation.

- Start a separate terminal and navigate to the application folder.

- Verify that the containers are running and copy the Conteiner ID of the application: 
```
sudo ps
```
- Connect to the application container by running the command (depending on OS) with the current containerId: 
```
docker exec -it <containerId> bash
```
or
```
docker exec -it <containerId> sh 
```
- Run the commands:
```
npx prisma generate
npx prisma migrate dev --name init
```
## Testing

- Test execution in the application container:
```
npm run test:auth
```
![image](https://github.com/hi-bi/nodejs2023Q2-service/assets/49794129/1c295cd3-8c6f-409a-9728-6f092c14d909)

- Command to clear data:
```
npx prisma db seed
```
- Check log level
  
  Change LOG_LEVEL in `.env`.

  Log levels:
  LOG = 0
  ERROR = 1
  WARN = 2
  DEBUG = 3
  VERBOSE = 4

- Check log to file(s)
  
  Set IS_FILE_LOGGING to `true`.
  Set IS_ERROR_FILE_LOGGING to `true` to log errors to separate file.
  If IS_FILE_LOGGING setted to `false`, errors will be output to the common `process.stdout`. 

  The LOGGING_GENERAL_FILE_SIZE and LOGGING_ERROR_FILE_SIZE define the file size in Kb.



# 
Please note that the code contains extra code, commented code and warnings for the assignment, which is left to be used for further investigation and development.  

If you have any questions, write to the Discord.
